man/man1/fracplanet.1
.TH FRACPLANET 1 "26 Apr 2006" "www.timday.com" "Fracplanet"

.SH NAME
fracplanet \- Create and view random fractal terrain and planets.

.SH SYNOPSIS
fracplanet
[
.I <QtOptions>
]
[
.B --help
]
[
.B --verbose
]
[
.B --display-list
]
[
.B --invert-mouse-y
]
[
.B --wireframe
]

.SH DESCRIPTION

.B Fracplanet 
generates random planets and terrain with oceans, 
mountains, icecaps and rivers.  Parameters are specified interactively 
and the results displayed using OpenGL.  The generated objects can be
exported for Pov-Ray or Blender.

A document fully describing usage via the application's GUI is
accessible from the application's "About" menu once it is running.

.SH GENERAL OPTIONS

.I <QtOptions>
Standard options recognised by the Qt toolkit (for example
.B \-geometry
.I <width>x<height>
Note that these use the single-dash option style.
).

.B --help, -h
Display a list of recognised arguments.

.B --verbose, -v
Display information about fracplanet and the GL graphics system it's running on.

.SH RENDERING OPTIONS
These affect the initial settings on the render controls tab, which controls the OpenGL display.

.B --display-list, -d
Start up the application with rendering in display list mode by default
(c.f immediate mode).
This is particularly useful when the application is running remotely as the
generated meshes are then only sent to the local OpenGL display hardware once.

.B --invert-mouse-y, -y
Start application with mouse in non-joystick mode for flight.

.B --wireframe, -w
Start application with rendering set to wireframe mode.

.SH SEE ALSO
.B blender
.B povray

.SH AUTHOR
.B fracplanet
was written by Tim Day (www.timday.com) and is released
under the conditions of the GNU General Public License.
For further details see the application's "About" dialog
or the file LICENSE supplied with the source code.



NEWS
From release 0.3.3:
 - Bump version number to 0.4.0 (0.4.0dev while under development)
 - Migrate sourceforge repository to svn
 - Port to Qt4
 - Precompiled headers
 - CHANGES renamed to NEWS
 - License boilerplate managed by "headache"

From release 0.3.2:
 - Bump version number to 0.3.3 (dev while under development)
 - Add menu entry
 - Update for etch & edgy (warning-free c++, package building)

From release 0.3.1:
 - Bump version number to 0.3.2 (dev while under development)
 - Reinstate weather systems (spinbox to control number).
 - Various commandline options (uses boost::program_options).
 - Save to texture, shaded and unshaded.
 - Illumination direction control

From release 0.3.0:
 - Bump version number to 0.3.1dev
 - Use boost's mt19937 for random number generation.
 - Minor GUI improvement (terrain type pull down).
 - Add simple man page.
 - Deb building script.

From release 0.2.0:
 - Bump version number to 0.3.0
 - Different flat terrain types 
 - Blender export (faux alpha for clouds).
 - Internal changes (RGB to RGBA, overload alpha for emission)
 - Built in docs
 - Cloud layer (with Blender and POV support).
 - Render (OpenGL) background colour control

From release 0.1.0:
 - Bump version number to 0.2.0
 - Progress dialog separated and better behaved
 - Variation uses % scale for consistency (was 1/1000ths before)
 - Cosmetic changes
 - Add multiscale Perlin noise step
 - River generation faster and more reliable.
 - Subdivision enumerated as 1...levels in progress dialog (not from zero)
 - Improvements to river generation memory usage.
 - Saved povray file no longer needs colors.inc

From release 0.0.3:
 - Bump version number to 0.1.0
 - Support emissive oceans and rivers (including POV save)
 - Fix .inc filename bug (spurious ")
 - Update docs
 - Viewer to separate top-level.
 - Don't burn CPU when OpenGL area not visible.
 - Crude flight-simulator mode.  Joystick-mouse selector.
 - Save uses Qt file-picker dialog
 - Slider control of ambient
 - Amount of movement independent of FPS rate.
 - Clean up bogus files in CVS

From release 0.0.1:
 - Bump version number to 0.0.3 (0.0.2 omitted docs)
 - Add fracplanet.htm, fracplanet.css user documentation.
 - Minor changes to code (compiles on RH9)
 - Tweak doxygen.cfg 
 - Display list option
 - Frame rate display
 - VERSION and mktgz done evolvotron style

From release 0.0.0:
 - CHANGES file added.
 - lake_becomes_sea initialisation changed from 5 to 0.05 (so displays "5" on controls instead of "100").
 - Remove .qmake.internal.cache from CVS control (don't you just HATE it when that happens)
 - Move hardwired colours into ParametersTerrain class (in anticipation of control from GUI).
 - Add colour picking to GUI.
 - Bump version number to 0.0.1 (appears only in control_about.cpp)
 - Generated .pov file correctly includes basename.inc (was always including terrain.inc before).



fracplanet.cpp
/**************************************************************************/
/*  Copyright 2009 Tim Day                                                */
/*                                                                        */
/*  This file is part of Fracplanet                                       */
/*                                                                        */
/*  Fracplanet is free software: you can redistribute it and/or modify    */
/*  it under the terms of the GNU General Public License as published by  */
/*  the Free Software Foundation, either version 3 of the License, or     */
/*  (at your option) any later version.                                   */
/*                                                                        */
/*  Fracplanet is distributed in the hope that it will be useful,         */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of        */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         */
/*  GNU General Public License for more details.                          */
/*                                                                        */
/*  You should have received a copy of the GNU General Public License     */
/*  along with Fracplanet.  If not, see <http://www.gnu.org/licenses/>.   */
/**************************************************************************/

/*! \mainpage Fracplanet : fractal terrain generator

  \author Tim Day 

  \section introduction Introduction
  "Fracplanet" is an interactive tool for generating fractal planets and terrains.
  It can output the generated meshes to files suitable for use by POV-Ray.

  \todo For new features to be added, see the TODO file.
 */

#include "precompiled.h"

#include "fracplanet_main.h"

//! Application code
/*! Currently this simply creates a TriangleMesh object of some sort,
  then passes it to a viewer.
 */
int main(int argc,char* argv[])
{
  QApplication app(argc,argv);

  boost::program_options::variables_map opts;
  try
    {
      boost::program_options::options_description opt_desc
	("Recognised options (besides Qt standards):");

      opt_desc.add_options()
	("help,h","show list of recognised options")
	("verbose,v","verbose output to stderr")
	;
	
      opt_desc.add(ParametersRender::options());

      boost::program_options::store
	(
	 boost::program_options::parse_command_line(argc,argv,opt_desc),
	 opts
	 );
      boost::program_options::notify(opts);

      if (opts.count("help"))
	{
	  std::cerr << opt_desc << std::endl;
	  return 1;
	}
    }
  catch (boost::program_options::error& e)
    {
      std::cerr << "Bad command line: " << e.what() << std::endl;
      std::cerr << "Use -h or --help to list recognised options" << std::endl;
      return 1;
    }

  const bool verbose=opts.count("verbose");

  if (verbose)
    std::cerr << "Setting up...\n";

  FracplanetMain*const main_widget=new FracplanetMain(0,&app,opts,verbose);

  if (verbose)
    std::cerr << "...setup completed\n";

  main_widget->show();

  // Do this after setup as that will have initialized GL
  if (verbose)
    {
      std::cerr << "Fracplanet:" << std::endl;
      std::cerr << "  sizeof(ByteRGBA) is " << sizeof(ByteRGBA) << " (4 is good)" << std::endl;  
      std::cerr << "  sizeof(Vertex)   is " << sizeof(Vertex) << " (32 is good)" << std::endl;
      std::cerr << "  sizeof(Triangle) is " << sizeof(Triangle) << " (12 is good)" << std::endl;
      std::cerr << std::endl;

      std::cerr << "GL:" << std::endl;

      std::cerr << "  Vendor   : " << glGetString(GL_VENDOR) << std::endl;
      std::cerr << "  Renderer : " << glGetString(GL_RENDERER) << std::endl;
      std::cerr << "  Version  : " << glGetString(GL_VERSION) << std::endl;

      GLint max_elements_vertices;
      GLint max_elements_indices; 
      glGetIntegerv(GL_MAX_ELEMENTS_VERTICES,&max_elements_vertices);
      glGetIntegerv(GL_MAX_ELEMENTS_INDICES,&max_elements_indices);      
      std::cerr << "  GL_MAX_ELEMENTS_VERTICES : " << max_elements_vertices << std::endl;
      std::cerr << "  GL_MAX_ELEMENTS_INDICES : " << max_elements_indices << std::endl;

      std::cerr << "  GL Extensions are : \"" << glGetString(GL_EXTENSIONS) << "\"" << std::endl;
      //std::cerr << "GLU Extensions are :\n\"" << gluGetString(GL_EXTENSIONS) << "\"\n";

      //std::cerr.flush();
    }

  std::cerr << "fracplanet: commencing main loop...\n";

  return app.exec();
}
